### DISCUSSION TITLE
Proposal for configuration model

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork3#4] [Ideas]
WW2 has these configuration files that execute in series:
`site.conf`, `defaults.config`, `localOverrides.conf`, `course.conf`, `simple.conf`
(I may have the first two out of order.)

There is also a config file for LTI that can be included anywhere in that chain.

For WW3, I propose a longer cascade of configuration. (Perhaps with files, perhaps all in the database—that's a separate question.)

After distribution defaults, site configuration, apache configuration...

1. Global overrides. "Global" with respect to everything on that server. This is like the current "localOverrides" (local wrt the larger commuity of WW servers.)
2. For lack of a better term "Region" configuration. Examples: within a school, you might have the math and chemistry department each using WW. They might have competing configuration needs. Courses could have an optional "region" attribute and if so, a region configuration is loaded. "Regions" might also be like "Calculus" versus "Statistics" where different configurations are desired. Or even like "MTH 111" versus "MTH 251". And then stepping back, regions could be used to manage when more than one institution uses one WW server, like MAA or Runestone, and set school theme configurations this way.
3. Course config. Like the current one.
4. Maybe something like the current course-wide simple.conf, but we could revisit whether that is needed.
5. Each problem set gets a configuration. For example, turning off MathQuill for one problem set, or turning on the display of walkthrough solutions for a formative problem set. This could include non-problem material like the current set header. This is less workable as a file, and would work better as something in the database.
6. Each _problem_ can be configured. Again, in the database.

**Comment by pstaabp** [comment]
This seems reasonable as long as it is clear on the cascading.  That is, what is the order and we should clearly say and test what overrides what.
